### PR TITLE
add continuation-local-storage (CLS) support

### PR DESCRIPTION
**Pull request by akhoury**
Ok hear me out.. you no longer need to plumb the `req` object through arguments, to anywhere, either for post create, or to a specific `filter/action` hook(s) so just you can have access to `req.ip` for example... all you have to do now, in ANY module, or ANY plugin is the following..

```javascript
	var cls = require('./middleware/cls'); // require this once in your module.
        // ..... 

       var onReply = function () {  // or whatever
             var req = cls.get('request'); 
             // this is THE req object for THAT web request in the calls chain, 
             // you can access, ip, session info from.. whatever
             console.log(req); // from here, sky is the limit.
        }
```

read more about this here, most likely will become in node core eventually, or at least a very similar solution, as this extra dependency is a [shim](https://github.com/othiym23/node-continuation-local-storage/blob/4b5cfc4f47d97611fc5238e18c8b5abf2d5bb669/context.js#L15) at the moment.

https://github.com/nodejs/node-v0.x-archive/issues/5243
https://github.com/nodejs/node-v0.x-archive/pull/6011
SHIM: https://github.com/othiym23/node-continuation-local-storage

I do not have any strong opinion on namespace name, storage key name.. or location of this extra `cls.js` .. please let me know what you think/feel about this.

I just hate the fact that, "sometimes" we plumb through the req object, and sometimes we don't, and when we do, it's not really part of its sibling properties in the same data hash, it's just there because we needed access to the IP address or whatever..

### CODE REVIEW

**Comment by julianlam** [comment]
> I just hate the fact that, "sometimes" we plumb through the req object, and sometimes we don't, and when we do, it's not really part of its sibling properties in the same data hash, it's just there because we needed access to the IP address or whatever..

I agree, it was an inelegant solution to begin with, but passed our checks because there was basically no other easy option. This works, and if it's on track to be included in Node.js core, then I have no objections at all.

**Comment by barisusakli** [comment]
How does this work in places where we don't have the req object, ie socket calls.

We retrieve the uid and ip from the socket object in those places.

**Comment by akhoury** [comment]
@barisusakli i could add handler for socket connection and expose the socket the same way, .. on connection style - never tested, but will do.

```
io.on('connection', function(socket) {
    storage.set('socket', socket)
})
```

**Comment by julianlam** [comment]
Awesome, yeah please give this a shot. It would make our code more elegant, without having to explicitly pass in `req`/`res`.

**Comment by akhoury** [comment]
added websockets support - so now.. anywhere in the App or any plugin, 

```javascript
var cls = require('/path/to/nodebb/src/middleware/cls'),

// .... 

onWhatever: function () {
    console.log( cls.get('http') ); // prints { req: req, res: res }
    console.log( cls.get('ws') ); // prints { socket: socket, payload: payload /* if available */, event: event }
}
```

**Comment by julianlam** [comment]
Whoever merges will also need to add deprecation notices to all plugin hooks currently passing in `req`,  socket, etc.

**Comment by akhoury** [comment]
i can add those first - and maybe some more documentation, where are the docs maintained ?

**Comment by psychobunny** [comment]
This is really awesome :+1:

**Comment by psychobunny** [comment]
You can edit the docs right on GH, there's a link on every page that takes you to the edit page. There's a hook set up that regenerates the docs after every change

File: src/plugins/hooks.js [hunk 1]
```diff
@@ -71,6 +74,20 @@ module.exports = function(Plugins) {
...
+				&& params
+				&& Array.isArray(hookList)
+				&& hookList.length
+				&& (params.req || params.res || params.socket || params.uid)) {
+
+			Plugins.clsDeprecatedParamsWarnedHooks[hook] = true;
+
+			winston.warn('[plugins] hook `' + hook + '` \'s `params.req`, `params.res`, `params.uid` and `params.socket` are being deprecated, '
+			+ 'plugins should use the `middleware/cls` module instead to get a reference to the http-req/res and socket (which you can get the current `uid`) '
+			+ '- for more info, visit https://docs.nodebb.org/en/latest/plugins/create.html#getting-a-reference-to-req-res-socket-and-uid-within-any-plugin-hook');
```
Comment (akhoury) [inline src/plugins/hooks.js]: that link here doesn't work yet, it's the expected link to be after you push [nodebb-english](https://github.com/NodeBB/nodebb-english) live.

File: src/plugins/hooks.js [hunk 2]
```diff
@@ -71,6 +74,20 @@ module.exports = function(Plugins) {
 		var hookList = Plugins.loadedHooks[hook];
 		var hookType = hook.split(':')[0];
 
+		// todo: remove when breaking all hooks params by removing req/res/socket/uid
+		if (!Plugins.clsDeprecatedParamsWarnedHooks[hook]
```
Comment (akhoury) [inline src/plugins/hooks.js]: i really didn't want to touch over 30 files that fireHooks with the re

File: src/plugins/hooks.js (hunk 1 above)
Comment (akhoury) [inline src/plugins/hooks.js]: Also, please feel free to reword that long sentence or whatever.. i am not good at that stuff.

**Comment by akhoury** [comment]
let me know what you think about this solution for the deprecation warnings - there are too many files that needs to be touched otherwise - or we can do it on hook registrations but we'd have to make a list of them instead - which i could do i guess.

**Comment by akhoury** [comment]
i guess the silence means I have to manually create a list of hook

**Comment by akhoury** [comment]
I did a search for all instances of  `fireHook` and manually build this map of hooks, if I skipped anything please let me know or feel free to add it your self.

I think we should leave the `{ uid }` parameter untouched, meaning limit this CLS thing to, `req`, `res` and `socket`, even if hooks passes the `uid` around, I think that's fine, because it's part of the relevant data, especially if you have interaction between 2 users, for example, `follow/unfollow`, or `message`, 

I i guess, only `req`, `res` and `socket` should now be accessed via the CLS

Another note, for example, `filter:middleware.buildHeader` hook passes `{req: req, locals: res.locals}` as argument, I think `locals` should not be removed, because it's the relevant data to that hook here, even though we can access it from `res.local`. But `req` should be removed since it's not directly relevant.

**Comment by julianlam** [comment]
Okay, so what would this mean for plugins that currently subscribe to a hook that sent in these values that will be removed?

**Comment by akhoury** [comment]
This means plugins should not expect to get a reference of the `req`, `res` or `socket` references from the hooks' arguments anymore, but rather from the CLS module, as explained here: https://docs.nodebb.org/en/latest/plugins/create.html#getting-a-reference-to-req-res-socket-and-uid-within-any-plugin-hook

As far as the nodebb deprecation goes, we could this in 2 ways, although I wouldn't make this breaking change for 1.0, maybe an 1.x or even 2.0

# 1
if you merge, we could go around the `src/` and remove ALL the parameters of `req`, `res` and `socket` (although I didn't see any `socket`) of ALL hooks arguments, then inside `fireHook()` function, we can add them back to support backward compatibility for the plugins but keep the deprecation message until you guys decide to make the breaking change - then it would be consolidated in 1 place, in about 10 lines of code to remove then.

something like: 
```javascript
var cls = require('../middleware/cls');

// ....
Plugins.fireHook = function(hook, params, callback) {
		// ......
            
               if ( isHookParamsDeprecated(hook) ) {
                      params.req = cls.get('http').req;
                      params.res = cls.get('http').res;
                      params.socket = cls.get('ws').socket;
                }

                switch (hookType) {
                 ....
                }
	};
```

# 2
if you merge, then you do nothing extra, until you decide to make the breaking change one day, then on that day, we do step 1, and remove all the `req`, `res` parameters

**Comment by julianlam** [comment]
Okay, good to merge in that case, as long as it is not a breaking change upon merge. More than willing to use deprecation notices for all `1.x.x` releases, and finally deprecate in `2.x.x`

This can be merged post v1.0.0 however.

**Comment by julianlam** [comment]
By the way, what happens when you call `cls.get('ws')` when you're not in a "socketey" environment?

**Comment by akhoury** [comment]
> what happens when you call cls.get('ws') when you're not in a "socketey" environment

you get `undefined` - it's just a in-memory key/value pair store, but we should just protect against that as well, `(cls.get('ws') || {}).socket`

That is a very rare case IMO, because on any initial connection, `ws` will be set, the hook will need to execute on the server, and not be triggered by anything except something like a cron-job for example, because if, ANYWHERE in the calls stack (sync or async) - the hook was triggered via web-request, or a socket-event, `http` and/or `ws` will be set, even if it was just a http-request, the `ws` will be set as soon as the client connects.

**Comment by akhoury** [comment]
> This can be merged post v1.0.0 however

true, and I agree, it's better to release 1.0 first and pave the way for that change. Would you want me to close and re-submit after 1.0 or leave this PR open, and I will keep it in-sync with master. Maybe, while I'm at it, do the clean-up step 1 talked about [2 comments ago](https://github.com/NodeBB/NodeBB/pull/4162#issuecomment-190236035)

**Comment by julianlam** [comment]
Leave it PR'd against `master`. v1.0.0 launched today, so we can merge this in soon:tm:

**Comment by akhoury** [comment]
ok

**Comment by barisusakli** [comment]
I have one concern about this and its the `ws` vs `http` issue. This requires the plugin author to know from which context the hook is being called. In some cases it might even be called from both. 

Can we somehow do `cls('request').req` and if we are in http it returns the req object and if we are in socket land it returns `reqFromSocket(socket)` ? This way plugins don't have to check for req and socket in every hook.

**Comment by julianlam** [comment]
@barisusakli but even then, the object keys are different... in socket land, it's `.uid` to get the uid, but it's `.user.uid` in http land...

**Comment by barisusakli** [comment]
Nah we have req.uid for a while

**Comment by julianlam** [comment]
okay, that's just off the top of my head, but I think the problem still
stands?

I'd be in favour of a couple bools for quick determination:

```
{
    "isSocket": true,
   " isHttp": false
}
```

**Comment by barisusakli** [comment]
Well that doesn't solve my problem. Now inside plugins in every hook I would have to check 

```
if (isSocket) {
} else {
}
```

**Comment by julianlam** [comment]
Is it really that big of an issue? Http methods are registered explicitly via plugin.json, and websocket methods are added _strictly_ from within the plugin itself, so if anything, the plugin should handle cases where it's methods are used as both http and socket handlers?

**Comment by barisusakli** [comment]
{"hook": "filter:topic.post", "method": "checkReply", "priority": 5},
{"hook": "filter:topic.reply", "method": "checkReply", "priority": 5},

These are both fired from socket methods, but they could have been fired from http land as well.
So no they are not strictly added in the plugin. And its perfectly possible someone might call topics.reply from a http route or socket method.

**Comment by julianlam** [comment]
Okay, I understand the need for future proofing this... though I am at a loss for a solution at this time.

**Comment by barisusakli** [comment]
```
Sockets.reqFromSocket = function(socket) {
	var headers = socket.request.headers;
	var host = headers.host;
	var referer = headers.referer || '';

	return {
		ip: headers['x-forwarded-for'] || socket.ip,
		host: host,
		protocol: socket.request.connection.encrypted ? 'https' : 'http',
		secure: !!socket.request.connection.encrypted,
		url: referer,
		path: referer.substr(referer.indexOf(host) + host.length),
		headers: headers
	};
};
```

if we return the same req object for both http and socket.io land we should be good.

**Comment by akhoury** [comment]
@barisusakli I'm with you on that one. 

I like the idea of unifying the storage, however I think that we should still keep a reference on the socket data on the hash returned from `reqFromSocket`.. along with adding the payload (as `body`) and the `uid`

```javascript
Sockets.reqFromSocket = function(socket, payload, event) {
    var headers = socket.request.headers;
    var host = headers.host;
    var referer = headers.referer || '';

    // use the payload and event to add a body to the request?
    var body = payload || {};
    body.data = body.data || [];
    body.event = event || body.data[0];

return return {
        ip: headers['x-forwarded-for'] || socket.ip,
        host: host,
        protocol: socket.request.connection.encrypted ? 'https' : 'http',
        secure: !!socket.request.connection.encrypted,
        url: referer,
        path: referer.substr(referer.indexOf(host) + host.length),
        headers: headers,

        _socket: socket // <------ 
        uid: socket.uid, // didn't see the uid here 
        body: body
    };
```

## Question
Also, do you think there is any need  for the `res` object to be available? `storage.get('response')`

**Comment by barisusakli** [comment]
I don't think I ever needed `res` in my plugins not sure about the others. The most needed stuff is `req.uid/socket.uid` and `req.ip/socket.ip`

If `storage.get('request');` returns the same unified object in both contexts I think this is a good start.

**Comment by julianlam** [comment]
Plugins could set items in res.locals... I can't think of a current usage, but it is possible.

**Comment by akhoury** [comment]
> Plugins could set items in res.locals..

We could expose `storage.get('response')` also, it would just return undefined in sockety environment.

**Comment by akhoury** [comment]
Ok I think this is ready for a re-review. The socket.io/index.js file might look like it was changed a lot, but it hasn't, i had to merge the latest here and I had to fix the circular dependency with CLS, so the tabbing/spaces will throw you off a bit. Sorry about that. 


[usage](https://github.com/akhoury/nodebb-english/blob/patch-3/plugins/create.rst#getting-a-reference-to-each-request-from-within-any-plugin-hook)
```
var cls = module.parent.require('./middleware/cls'); // require cls once in your plugin.

var MyPlugin = {
        myMethod: function(postData, callback) {
            var request = cls.get('request'); // current http request object.
            var uid = request.uid; // current user id
            // ...
        },

        // let's say this one occurs on a websocket event,
        myOtherMethod: function(somethingData) {
            var request = cls.get('request'); // current returned from Sockets.reqFromSocket() object

            var uid = request.uid; // current user id, if available

            var payload = request.body; // socket payload data, if available
            var event = request.method; // socket last event, if available, not sure if using method is cool here
            var params = request.params; // socket last params, if available

            // ...
        }
    };

module.exports = MyPlugin;
```

**Comment by barisusakli** [comment]
Hmm, so what happens if a lot of socket events are coming in in a short period of time? Due to things being async I think there is a possibly that `cls.socket()` gets called twice or more before the function hook calls `cls.get` which would result in wrong body/method/params being reported in the hook method right?

From what I understand cls only stores one instance of the data.

**Comment by akhoury** [comment]
It won't matter, if you `cls.get()` within each request-call-stack*, each `get()` will get the corresponding data of whatever was saved somewhere up the request-call-stack.

> cls only stores one instance of the data.

true, but per each request-call-stack,

* the reason I call it the request-call-stack, but the usual call-stack we're used to, does not include timeouts and async calls, but cls handles all that. so if

```javascript
// on some http or socket request.
doSomething() {
    var r1 = cls.get('request');

     setTimeout(function() { 
        // this timeout callback will not be on same JS call stack as doSomething, but it would be on the same request-call-stack

        var r2 = cls.get('request');
        console.log(r1 === r2); // true
    }, 100)
}

**Comment by julianlam** [comment]
@psychobunny signed off [earlier](https://github.com/NodeBB/NodeBB/pull/4162#issuecomment-183978904)
I'll sign off on this now.

Once @barisusakli approves, this is good to merge.

**Comment by akhoury** [comment]
did i break the world?

**Comment by julianlam** [comment]
You're good :)

Julian Lam

Co-Founder, NodeBB Inc.
619-341 King Street East
Toronto, Ontario
M5A 1L1

**Pull request merged**